File: examples/experimental/agent_builder/customer_care_style/toolkits/banking_mcp_server/py_server/src/mortgage_service.py
"""
Mortgage Service - Mock Implementation

Contains mock business logic for mortgage operations.
"""

from dataclasses import dataclass
from datetime import datetime, timedelta
import random
import string


@dataclass
class MortgageBalance:
    """Mortgage balance information."""

    loan_number: str
    original_amount: float
    outstanding_balance: float
    interest_rate: float
    monthly_payment: float
    next_payment_date: str
    remaining_term_months: int


@dataclass
class MortgagePayment:
    """Mortgage payment record."""

    payment_date: str
    amount: float
    principal: float
    interest: float
    escrow: float
    balance_after: float


@dataclass
class PreparedPayment:
    """Prepared payment awaiting confirmation."""

    transaction_id: str
    customer_id: str
    loan_number: str
    amount: float
    payment_date: str
    current_balance: float
    expires_at: str


@dataclass
class PaymentValidationResult:
    """Result of payment date validation."""

    is_valid: bool
    error_message: str | None = None


class MortgageService:
    """Mock Mortgage Service."""

    @staticmethod
    def get_mortgage_balance(customer_id: str) -> MortgageBalance:
        """Get mortgage balance and details."""
        return MortgageBalance(
            loan_number=f"MTG-{customer_id}-001",
            original_amount=350000.00,
            outstanding_balance=287450.32,
            interest_rate=3.75,
            monthly_payment=1620.50,
            next_payment_date="2026-01-01",
            remaining_term_months=276,
        )

    @staticmethod
    def get_mortgage_payments(customer_id: str, count: int = 6) -> dict:
        """Get mortgage payment history."""
        loan_number = f"MTG-{customer_id}-001"
        all_payments = [
            MortgagePayment(
                payment_date="2025-12-01",
                amount=1620.50,
                principal=722.15,
                interest=898.35,
                escrow=0,
                balance_after=287450.32,
            ),
            MortgagePayment(
                payment_date="2025-11-01",
                amount=1620.50,
                principal=719.88,
                interest=900.62,
                escrow=0,
                balance_after=288172.47,
            ),
            MortgagePayment(
                payment_date="2025-10-01",
                amount=1620.50,
                principal=717.61,
                interest=902.89,
                escrow=0,
                balance_after=288892.35,
            ),
            MortgagePayment(
                payment_date="2025-09-01",
                amount=1620.50,
                principal=715.35,
                interest=905.15,
                escrow=0,
                balance_after=289609.96,
            ),
            MortgagePayment(
                payment_date="2025-08-01",
                amount=1620.50,
                principal=713.10,
                interest=907.40,
                escrow=0,
                balance_after=290325.31,
            ),
            MortgagePayment(
                payment_date="2025-07-01",
                amount=1620.50,
                principal=710.86,
                interest=909.64,
                escrow=0,
                balance_after=291038.41,
            ),
        ]

        return {
            "customer_id": customer_id,
            "loan_number": loan_number,
            "payments": [
                {
                    "payment_date": p.payment_date,
                    "amount": p.amount,
                    "principal": p.principal,
                    "interest": p.interest,
                    "escrow": p.escrow,
                    "balance_after": p.balance_after,
                }
                for p in all_payments[:count]
            ],
        }

    @staticmethod
    def validate_payment_date(payment_date: str) -> PaymentValidationResult:
        """Validate payment date."""
        today = datetime.now().replace(hour=0, minute=0, second=0, microsecond=0)
        try:
            payment_date_obj = datetime.strptime(payment_date, "%Y-%m-%d")
        except ValueError:
            return PaymentValidationResult(
                is_valid=False,
                error_message=f"Invalid date format: {payment_date}. Please use YYYY-MM-DD format.",
            )

        thirty_days_from_now = today + timedelta(days=30)

        # Check if date is in the past
        if payment_date_obj < today:
            return PaymentValidationResult(
                is_valid=False,
                error_message=(
                    f"We're unable to schedule a payment for {payment_date} as this date has "
                    "already passed. Please select today's date or a date within the next 30 days."
                ),
            )

        # Check if date is more than 30 days in the future
        if payment_date_obj > thirty_days_from_now:
            latest_date = thirty_days_from_now.strftime("%Y-%m-%d")
            return PaymentValidationResult(
                is_valid=False,
                error_message=(
                    f"We can only schedule payments up to 30 days in advance. The date you "
                    f"selected ({payment_date}) is beyond this window. Please choose a date "
                    f"on or before {latest_date}."
                ),
            )

        return PaymentValidationResult(is_valid=True)

    @staticmethod
    def prepare_payment(
        customer_id: str,
        amount: float,
        payment_date: str,
    ) -> PreparedPayment:
        """Prepare a mortgage payment."""
        loan_number = f"MTG-{customer_id}-001"
        current_balance = 287450.32
        random_suffix = "".join(random.choices(string.ascii_lowercase + string.digits, k=9))
        transaction_id = f"TXN-{int(datetime.now().timestamp() * 1000)}-{random_suffix}"
        # 5 minutes from now
        expires_at = (datetime.now() + timedelta(minutes=5)).isoformat()

        return PreparedPayment(
            transaction_id=transaction_id,
            customer_id=customer_id,
            loan_number=loan_number,
            amount=amount,
            payment_date=payment_date,
            current_balance=current_balance,
            expires_at=expires_at,
        )

    @staticmethod
    def process_mortgage_payment(
        customer_id: str,
        loan_number: str,
        amount: float,
        payment_date: str,
    ) -> str:
        """
        Process a mortgage payment.
        In a real implementation, this would call your mortgage system to process the payment.
        """
        # Mock implementation - in production this would:
        # - Process the payment with your mortgage system
        # - Update the mortgage balance
        # - Create payment records
        # - Return confirmation number from mortgage system
        random_suffix = "".join(random.choices(string.ascii_lowercase + string.digits, k=9))
        return f"CONF-{int(datetime.now().timestamp() * 1000)}-{random_suffix}"

File: examples/experimental/agent_builder/customer_care_style/toolkits/banking_mcp_server/ts_server/src/mortgageService.ts
/**
 * Mortgage Service - Mock Implementation
 *
 * Contains mock business logic for mortgage operations
 */

export interface MortgageBalance {
  loanNumber: string;
  originalAmount: number;
  outstandingBalance: number;
  interestRate: number;
  monthlyPayment: number;
  nextPaymentDate: string;
  remainingTermMonths: number;
}

export interface MortgagePayment {
  paymentDate: string;
  amount: number;
  principal: number;
  interest: number;
  escrow: number;
  balanceAfter: number;
}

export interface MortgagePaymentHistory {
  customerId: string;
  loanNumber: string;
  payments: MortgagePayment[];
}

export interface PreparedPayment {
  transactionId: string;
  loanNumber: string;
  amount: number;
  paymentDate: string;
  currentBalance: number;
  expiresAt: string;
}

export interface PaymentValidationError {
  isValid: false;
  errorMessage: string;
}

export interface PaymentValidationSuccess {
  isValid: true;
}

export type PaymentValidationResult =
  | PaymentValidationError
  | PaymentValidationSuccess;

/**
 * Mock Mortgage Service
 */
export class MortgageService {
  /**
   * Get mortgage balance and details
   */
  static getMortgageBalance(customerId: string): MortgageBalance {
    return {
      loanNumber: `MTG-${customerId}-001`,
      originalAmount: 350000.0,
      outstandingBalance: 287450.32,
      interestRate: 3.75,
      monthlyPayment: 1620.5,
      nextPaymentDate: '2026-01-01',
      remainingTermMonths: 276,
    };
  }

  /**
   * Get mortgage payment history
   */
  static getMortgagePayments(
    customerId: string,
    count: number = 6,
  ): MortgagePaymentHistory {
    const loanNumber = `MTG-${customerId}-001`;
    const allPayments: MortgagePayment[] = [
      {
        paymentDate: '2025-12-01',
        amount: 1620.5,
        principal: 722.15,
        interest: 898.35,
        escrow: 0,
        balanceAfter: 287450.32,
      },
      {
        paymentDate: '2025-11-01',
        amount: 1620.5,
        principal: 719.88,
        interest: 900.62,
        escrow: 0,
        balanceAfter: 288172.47,
      },
      {
        paymentDate: '2025-10-01',
        amount: 1620.5,
        principal: 717.61,
        interest: 902.89,
        escrow: 0,
        balanceAfter: 288892.35,
      },
      {
        paymentDate: '2025-09-01',
        amount: 1620.5,
        principal: 715.35,
        interest: 905.15,
        escrow: 0,
        balanceAfter: 289609.96,
      },
      {
        paymentDate: '2025-08-01',
        amount: 1620.5,
        principal: 713.1,
        interest: 907.4,
        escrow: 0,
        balanceAfter: 290325.31,
      },
      {
        paymentDate: '2025-07-01',
        amount: 1620.5,
        principal: 710.86,
        interest: 909.64,
        escrow: 0,
        balanceAfter: 291038.41,
      },
    ];

    return {
      customerId,
      loanNumber,
      payments: allPayments.slice(0, count),
    };
  }

  /**
   * Validate payment date
   */
  static validatePaymentDate(paymentDate: string): PaymentValidationResult {
    const today = new Date();
    today.setHours(0, 0, 0, 0);
    const paymentDateObj = new Date(paymentDate);
    paymentDateObj.setHours(0, 0, 0, 0);
    const thirtyDaysFromNow = new Date(today);
    thirtyDaysFromNow.setDate(thirtyDaysFromNow.getDate() + 30);

    // Check if date is in the past
    if (paymentDateObj < today) {
      return {
        isValid: false,
        errorMessage: `We're unable to schedule a payment for ${paymentDate} as this date has already passed. Please select today's date or a date within the next 30 days.`,
      };
    }

    // Check if date is more than 30 days in the future
    if (paymentDateObj > thirtyDaysFromNow) {
      const latestDate = thirtyDaysFromNow.toISOString().split('T')[0];
      return {
        isValid: false,
        errorMessage: `We can only schedule payments up to 30 days in advance. The date you selected (${paymentDate}) is beyond this window. Please choose a date on or before ${latestDate}.`,
      };
    }

    return { isValid: true };
  }

  /**
   * Prepare a mortgage payment
   */
  static preparePayment(
    customerId: string,
    amount: number,
    paymentDate: string,
  ): PreparedPayment {
    const loanNumber = `MTG-${customerId}-001`;
    const currentBalance = 287450.32;
    const transactionId = `TXN-${Date.now()}-${Math.random().toString(36).substr(2, 9)}`;
    const expiresAt = new Date(Date.now() + 5 * 60 * 1000).toISOString(); // 5 minutes from now

    return {
      transactionId,
      loanNumber,
      amount,
      paymentDate,
      currentBalance,
      expiresAt,
    };
  }

  /**
   * Get mock transaction for confirmation/cancellation
   * In a real implementation, this would look up the transaction from storage
   */
  static getMockTransaction(customerId: string) {
    return {
      loanNumber: `MTG-${customerId}-001`,
      amount: 1620.5,
      paymentDate: new Date().toISOString().split('T')[0],
    };
  }

  /**
   * Process a mortgage payment
   * In a real implementation, this would call your mortgage system to process the payment
   */
  static processMortgagePayment(
    customerId: string,
    loanNumber: string,
    amount: number,
    paymentDate: string,
  ): string {
    // Mock implementation - in production this would:
    // - Process the payment with your mortgage system
    // - Update the mortgage balance
    // - Create payment records
    // - Return confirmation number from mortgage system

    return `CONF-${Date.now()}-${Math.random().toString(36).substr(2, 9)}`;
  }
}
